### PR TITLE
NO-ISSUE Add a PULL_REQUEST_TEMPLATE for assisted-service

### PR DESCRIPTION
**Pull request by flaper87**
# Description

This template will be proposed every time a new PR is created. The
author of the PR will be prompted with it and expected to fill in the
suggested sections, as needed.

None of these sections are required but they are definitely recommended
as they will make the review process easier for everyone.

More on PR templates here: https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests

Signed-off-by: Flavio Percoco <flavio@redhat.com>

# What environments does this code impact?

- [ ] Cloud
- [ ] Onprem
- [x] None

# How was this code tested?

Please, select one or more if needed:

- [ ] Local unit-tests
- [ ] Local Kupe API subsystem tests
- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] No tests needed

It's recommended to take a few extra minutes to provide more information about
how this code was tested. Here are some questions that may be worth answering:

- Should this PR be tested by the reviewer?
- Is this PR relying on CI for an e2e test run?
- Should this PR be tested in a specific environment?
- Any logs, screenshots, etc that can help with the review process?

# Assignees

Please, add one or two reviewers that could help review this PR.

/assign @celebdor 
/assign @ronniel1 
/assign @romfreiman 
/assign @filanov 
/assign @YuviGold 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated.
- [x] Reviewers have been listed
- [ ] This change does not require a documentation update

### CODE REVIEW

**Comment by filanov** [comment]
You can probably add another section in the main readme related to contributors and link the new doc

File: docs/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+# What environments does this code impact?
+
+- [ ] Cloud
+- [ ] Onprem
+- [x] None
+
+
+# How was this code tested?
+
+Please, select one or more if needed:
```
Comment (filanov) [inline docs/pull_request_template.md:22]: In case of code changes we should mention that we require unit-tests. 
Maybe need to add guidelines to the reviewers as well.

File: docs/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+You can refer to [Kubernete's community documentation] on writing good commit messages, which provides good tips and ideas.
+
+Some PRs addess specific issues. Please, refer to the [CONTRIBUTING] documentation for more
+information on how to link a PR to an existing issue.
+
+[Kubernete's community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
+[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md
+
+# What environments does this code impact?
```
Comment (filanov) [inline docs/pull_request_template.md:13]: Maybe you can add controller/operator

File: docs/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+Please, add one or two reviewers that could help review this PR.
+
+/assign @
+/assign @
+
+## Checklist
+
+- [ ] Title and description added to both, commit and PR.
+- [ ] Relevant issues have been associated.
+- [ ] Reviewers have been listed
```
Comment (YuviGold) [inline docs/pull_request_template.md:52]: This part is automatic

File: docs/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+- Should this PR be tested in a specific environment?
+- Any logs, screenshots, etc that can help with the review process?
+
+
+# Assignees
+
+Please, add one or two reviewers that could help review this PR.
+
+/assign @
+/assign @
```
Comment (YuviGold) [inline docs/pull_request_template.md:46]: This part is automatic

File: docs/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+# How was this code tested?
+
+Please, select one or more if needed:
+
+- [ ] Local unit-tests
+- [ ] Local Kupe API subsystem tests
+- [ ] assisted-test-infra environment
+- [ ] dev-scripts environment
+- [ ] Reviewer's test appreciated
```
[the remark this answers is not in the document]
Comment (flaper87) [reply docs/pull_request_template.md:28]: > I would never expect a reviewer to run tests. The CI should always cover that

I agree that CI should always cover that but, realistically, this is not always the case. Flagging this doesn't mean the reviewer *has* to run the tests. But there are cases where this may be valuable.

I have tested other ppl's PRs in my environments and that has proven to be useful when CI is not covering the work being done.

File: docs/pull_request_template.md (hunk 4 above)
Comment (flaper87) [reply docs/pull_request_template.md:46]: I would like for contributors to think about who could help with their reviews. The reviewers picked are not always the right/best ones.

I think having this, at least until we clean-up the repo a bit and create proper OWNERS files for every subpath, is a good exercise.

File: docs/pull_request_template.md [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+Please, add one or two reviewers that could help review this PR.
+
+/assign @
+/assign @
+
+## Checklist
+
+- [ ] Title and description added to both, commit and PR.
+- [ ] Relevant issues have been associated.
```
[the remark this answers is not in the document]
Comment (flaper87) [reply docs/pull_request_template.md:51]: > as part of the title of the PR / summary of the commit

Not sure if you are proposing to add this to the check-list or sharing it as a motivation to remove this point for the list.

File: docs/pull_request_template.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+
+# How was this code tested?
+
+Please, select one or more if needed:
+
+- [ ] Local unit-tests
+- [ ] Local Kupe API subsystem tests
+- [ ] assisted-test-infra environment
+- [ ] dev-scripts environment
```
[the remark this answers is not in the document]
Comment (flaper87) [reply docs/pull_request_template.md:27]: I would like this list to be explicit about the environment where the PR was tested. `e2e workflow` doesn't say much to me. 

If you think `e2e workflow` covers something not listed here, then we can add it but I would prefer `dev-scripts environment` to stay in this list

File: docs/pull_request_template.md (hunk 1 above)
Comment (flaper87) [reply docs/pull_request_template.md:22]: Indeed! 

I will mention this and I will also add a `Reviewers checklist` to this template. Eventually, once we have the guidelines docs, we can also link them here.

**Review by flaper87 — COMMENTED** [review]
@filanov @YuviGold thanks for the initial review. I like where this is going. Will wait for some more feedback before addressing your comments :pray:

**Comment by flaper87** [comment]
> @filanov @YuviGold thanks for the initial review. I like where this is going. Will wait for some more feedback before addressing your comments

I have addressed your comments

File: docs/pull_request_template.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+- [ ] Operator
+- [x] None
+
+
+# How was this code tested?
+
+Please, select one or more if needed:
+
+- [ ] Local unit-tests
+- [ ] Local Kupe API subsystem tests
```
Comment (empovit) [inline docs/pull_request_template.md:24]: Misspelled Kube API?

File: docs/pull_request_template.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (empovit) [reply docs/pull_request_template.md:27]: _How was this code tested?_ Is "manually" an option? Sometimes it's not possible/feasible to write automated tests. In such cases maybe we should let the developer describe a procedure for testing the particular.

File: docs/pull_request_template.md (hunk 4 above)
Comment (empovit) [reply docs/pull_request_template.md:45]: @YuviGold We already had a discussion about this a while ago. Automatic assignments create too much noise - the bot can't take into consideration the areas of expertise of a developer, so sometimes the assigned reviewer(s) can't do a meaningful review, or need a lot of time just to understand the context. 

If it's not possible to disable the automatic assignment, we at least should carefully /cc and /uncc, IMO.

File: docs/pull_request_template.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+/assign @
+/assign @
+
+## Checklist
+
+- [ ] Title and description added to both, commit and PR.
+- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
+- [ ] Reviewers have been listed
+- [ ] This change does not require a documentation update
```
Comment (empovit) [inline docs/pull_request_template.md:52]: Documentation? Meaning README and such? Godoc/Python Docstring for APIs?

File: docs/pull_request_template.md (hunk 8 above)
Comment (flaper87) [reply docs/pull_request_template.md:24]: Nope, I have changed it forever!

File: docs/pull_request_template.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (flaper87) [reply docs/pull_request_template.md:27]: > How was this code tested? Is "manually" an option? Sometimes it's not possible/feasible to write automated tests. In such cases maybe we should let the developer describe a procedure for testing the particular.

I like this. To me, dev-script / test-infra is already a manual test but I see where you are coming from. Adding manually may encourage people to elaborate on how they did it.

File: docs/pull_request_template.md (hunk 9 above)
Comment (flaper87) [reply docs/pull_request_template.md:52]: All of it, I will be explicit

File: docs/pull_request_template.md (hunk 8 above)
Comment (empovit) [reply docs/pull_request_template.md:24]: Kute API :)

File: docs/pull_request_template.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+You can refer to [Kubernete's community documentation] on writing good commit messages, which provides good tips and ideas.
+
+Some PRs addess specific issues. Please, refer to the [CONTRIBUTING] documentation for more
+information on how to link a PR to an existing issue.
+
+# What environments does this code impact?
+
+- [ ] Cloud
+- [ ] Onprem
```
Comment (carbonin) [inline docs/pull_request_template.md:13]: Onprem is a bit of a tough term. What do we mean by "onprem" if we also have "operator" here?
Also "operator" and "controller" could be confusing. Maybe we need some comments in the template that explain when these would need to be checked?

For example, do I check "Operator" only when I'm changing the agent service config controller or would I use it also when changing local auth (which is only used currently by the serice when deployed with the operator)?

File: docs/pull_request_template.md [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+- [ ] Does this change includes unit-tests (note that code changes require unit-tests)
+
+## Reviewers Checklist
+
+- [ ] Are the title and description (in both PR and commit) meaningful and clear?
+- [ ] Is there a bug required (and linked) for this change?
+- [ ] Should this PR be backported?
+
+
+[Kubernete's community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
```
Comment (carbonin) [inline docs/pull_request_template.md:63]: The `'` is not needed here.
Even if it was possessive I think it would be `Kubernetes'`.

File: docs/pull_request_template.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
+# Description
+
+Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.
+
+You can refer to [Kubernete's community documentation] on writing good commit messages, which provides good tips and ideas.
```
Comment (carbonin) [inline docs/pull_request_template.md:5]: No apostrophe here.

File: docs/pull_request_template.md (hunk 8 above)
Comment (carbonin) [reply docs/pull_request_template.md:24]: Also on this line, but unrelated to spelling is that we have subsystem tests that are not for the kube-api.
I don't think those are covered by any of these bullets (unless you include that in `unit-tests`, which I wouldn't expect).

Maybe just drop the `Kube/Kube API` here entirely and leave it as just subsystem?

Comment (flaper87) [reply docs/pull_request_template.md:24]: > Maybe just drop the Kupe/Kube API here entirely and leave it as just subsystem?

:+1:

File: docs/pull_request_template.md (hunk 10 above)
Comment (flaper87) [reply docs/pull_request_template.md:13]: > For example, do I check "Operator" only when I'm changing the agent service config controller or would I use it also when changing local auth (which is only used currently by the service when deployed with the operator)?

You make a good point. I will remove `onprem` as I had added it before adding `Controller` and `Operator` but I will leave the comments/docs on what each part means for a follow-up PR so that we can start moving this one along.

File: docs/pull_request_template.md [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+- [ ] Cloud
+- [ ] Operator Managed Deployments
+- [x] None
+
+
+# How was this code tested?
+
+Please, select one or more if needed:
+
+- [ ] subsystem tests
```
Comment (filanov) [inline docs/pull_request_template.md:21]: I think that if someone review he will see that subsystem was added a well, so i'm not sure if it's needed. 
I think that all this section is more related to the review guidelines and not part of the PR.

Comment (flaper87) [reply docs/pull_request_template.md:21]: I have intended this section as a way for the author to communicate how the PR was tested. In some cases subsystem is enough (in this case there is probably a subsystem test added but in some cases, the current tests may already cover the PR).

**Review by filanov — APPROVED** [review]
